source-snowflake/discovery.go

```diff
@@ -0,0 +1,383 @@
+	var primaryKeys []*snowflakeDiscoveryPrimaryKey
+	if err := xdb.Select(&primaryKeys, primaryKeysQuery); err != nil {
+		return nil, fmt.Errorf("error listing primaryKeys: %w", err)
+	}
+	return &snowflakeDiscoveryResults{
+		Tables:      tables,
+		Columns:     columns,
+		PrimaryKeys: primaryKeys,
+	}, nil
+}
+
+func recommendCollectionName(table snowflakeObject) string {
+	if table.Schema != "PUBLIC" {
+		return strings.ToLower(table.Schema + "_" + table.Name)
+	}
+	return strings.ToLower(table.Name)
+}
+
+type snowflakeDiscoveryInfo struct {
+	Table      *snowflakeDiscoveryTable
+	Columns    []*snowflakeDiscoveryColumn
+	PrimaryKey []string
+}
+
+func catalogFromDiscovery(cfg *config, info *snowflakeDiscoveryResults) ([]*pc.Response_Discovered_Binding, error) {
+	log.Debug("translating discovery results into bindings")
+
+	// Aggregate information about the available streams.
+	var streams = make(map[snowflakeObject]*snowflakeDiscoveryInfo)
+	for _, discoveredTable := range info.Tables {
+		if discoveredTable.Database != cfg.Database {
+			return nil, fmt.Errorf("internal error: discovery results from other databases (this should never happen)")
+		}
+		if discoveredTable.Schema == cfg.Advanced.FlowSchema {
+			// Ignore objects in the 'Flow' schema so we don't accidentally discover staging tables.
+			continue
+		}
+		var tableID = snowflakeObject{discoveredTable.Schema, discoveredTable.Name}
+		if !strings.EqualFold(discoveredTable.Kind, "TABLE") {
+			log.WithFields(log.Fields{"table": tableID.String(), "kind": discoveredTable.Kind}).Trace("ignoring non-table entity")
+			continue
+		}
+		streams[tableID] = &snowflakeDiscoveryInfo{Table: discoveredTable}
+	}
+	for _, discoveredColumn := range info.Columns {
+		if discoveredColumn.Database != cfg.Database {
+			return nil, fmt.Errorf("internal error: discovery results from other databases (this should never happen)")
+		}
+		var tableID = snowflakeObject{discoveredColumn.Schema, discoveredColumn.Table}
+		if streams[tableID] == nil {
+			continue
+		}
+		streams[tableID].Columns = append(streams[tableID].Columns, discoveredColumn)
+	}
+	for _, discoveredPK := range info.PrimaryKeys {
+		if discoveredPK.Database != cfg.Database {
+			return nil, fmt.Errorf("internal error: discovery results from other databases (this should never happen)")
+		}
+		var tableID = snowflakeObject{discoveredPK.Schema, discoveredPK.Table}
+		if streams[tableID] == nil {
+			continue
+		}
+		streams[tableID].PrimaryKey = append(streams[tableID].PrimaryKey, discoveredPK.ColumnName)
+		if len(streams[tableID].PrimaryKey) != discoveredPK.KeySeq {
+			log.WithFields(log.Fields{
+				"stream": tableID.String(),
+				"pkey":   streams[tableID].PrimaryKey,
+				"seq":    discoveredPK.KeySeq,
+			}).Warn("primary key sequence mismatch")
+			return nil, fmt.Errorf("internal error: primary key sequence mismatch")
```

Comment:
I'm not entirely sure why, but I hit this error when trying to discover a checkpoints table that our snowflake materialization makes, `FLOW_CHECKPOINTS_V1`. It's primary keys don't seem to show up in order when I use a `SHOW PRIMARY KEYS`: 
<img width="394" alt="image" src="https://github.com/estuary/connectors/assets/55118525/4f335c1f-db9b-46ea-939b-a65e107bc4b7">

Reply:
Oh dang, I don't think I actually tested multiple-column primary keys very much. This is an obvious result of using `SHOW PRIMARY KEYS` as the enumeration query, there's no actual `ORDER BY key_sequence` occurring at any point.

I'll have to go re-investigate whether there's any way to get the necessary info from `information_schema` or if we have to just cope with `SHOW PRIMARY KEYS` returning unordered results and sort them ourselves...

Reply:
Yep, I've just double-checked. The `information_schema.table_constraints` view has existed for at least five years and for that entire time people keep asking and being told that there is still no way to actually get the _column names_ corresponding to a table's primary-key constraint other than by running `SHOW PRIMARY KEYS`.

I'll have to modify the logic to handle sorting the results of that query within the connector.

Reply:
The necessary sorting has been added explicitly in the code (see https://github.com/estuary/connectors/pull/1137/files#diff-3a385a923b61bcbfd54148783038ddf3a99bc5ed2235a25246767b42425b9587R117)